core/network/hostport.go

```diff
@@ -13,208 +13,361 @@ import (
+// as a flat slice of HostPort, preserving the order.
+func CollapseToHostPorts(serversHostPorts []MachineHostPorts) HostPorts {
+	var collapsed HostPorts
+	for _, hps := range serversHostPorts {
+		for _, hp := range hps {
+			collapsed = append(collapsed, hp)
 		}
 	}
-	return hps
+	return collapsed
+}
+
+// ProviderHostPort associates a provider/space aware address with a port.
+type ProviderHostPort struct {
+	ProviderAddress
+	NetPort
 }
 
-// ParseHostPorts creates a list of HostPorts parsing each given
-// string containing address:port. An error is returned if any string
-// cannot be parsed as HostPort.
-func ParseHostPorts(hostPorts ...string) ([]HostPort, error) {
-	hps := make([]HostPort, len(hostPorts))
+var _ HostPort = ProviderHostPort{}
+
+// String implements Stringer.
+func (hp ProviderHostPort) String() string {
+	return DialAddress(hp)
+}
+
+// GoString implements fmt.GoStringer.
+func (hp ProviderHostPort) GoString() string {
+	return hp.String()
+}
+
+type ProviderHostPorts []ProviderHostPort
+
+func (hp ProviderHostPorts) Addresses() []ProviderAddress {
+	addrs := make([]ProviderAddress, len(hp))
+	for i, hp := range hp {
+		addrs[i] = hp.ProviderAddress
+	}
+	return addrs
+}
+
+func (hp ProviderHostPorts) HostPorts() HostPorts {
+	addrs := make(HostPorts, len(hp))
+	for i, hp := range hp {
+		addrs[i] = hp
+	}
+	return addrs
+}
+
+// ParseMachineHostPorts creates a slice of MachineHostPorts parsing each given
+// string containing address:port.
+// An error is returned if any string cannot be parsed as a MachineHostPort.
+func ParseProviderHostPorts(hostPorts ...string) (ProviderHostPorts, error) {
+	hps := make(ProviderHostPorts, len(hostPorts))
 	for i, hp := range hostPorts {
-		hostport, err := ParseHostPort(hp)
+		mhp, err := ParseMachineHostPort(hp)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		hps[i] = *hostport
+		hps[i] = ProviderHostPort{
+			ProviderAddress: ProviderAddress{MachineAddress: mhp.MachineAddress},
+			NetPort:         mhp.NetPort,
+		}
 	}
 	return hps, nil
 }
 
-// ParseHostPort converts a string containing a single host and port
-// value to a HostPort.
-func ParseHostPort(hp string) (*HostPort, error) {
-	host, port, err := net.SplitHostPort(hp)
-	if err != nil {
-		return nil, errors.Annotatef(err, "cannot parse %q as address:port", hp)
-	}
-	numPort, err := strconv.Atoi(port)
-	if err != nil {
-		return nil, errors.Annotatef(err, "cannot parse %q port", hp)
-	}
-	return &HostPort{
-		Address: NewAddress(host),
-		Port:    numPort,
-	}, nil
+// SpaceHostPort associates an address with a port.
+type SpaceHostPort struct {
+	SpaceAddress
+	NetPort
 }
 
-// HostsWithoutPort strips the port from each HostPort, returning just
-// the addresses.
-func HostsWithoutPort(hps []HostPort) []Address {
-	addrs := make([]Address, len(hps))
+var _ HostPort = SpaceHostPort{}
+
+// String implements Stringer.
+func (hp SpaceHostPort) String() string {
```

Comment:
Not ideal to re-declare these for every type, but it is a symptom of the method of embedding.
Since they embed the address types and do not derive from `MachineHostPort`, they need to be duplicated, or the `String` and `GoString` methods are hoisted from the embedded address.